File: COMP9041/lab06/all_whales.pl
#!/usr/bin/perl -w

@pod = ();
@counter = ();
@whale = ();
@result_list = ();

while (my $line = <STDIN>) {
	chomp($line);
	$line =~ s/ +/ /g;
	$line =~ s/ +$//g;
	@line_split = split / /, $line, 2;
	$number = $line_split[0];
	$key_raw = $line_split[1];

	$key_raw_lc = lc $key_raw;
	if ($key_raw_lc =~ /.+s$/) {
		$key_raw_lc_singular = substr($key_raw_lc, 0, -1);
		$key = $key_raw_lc_singular;
	}
	else {
		$key = $key_raw_lc;
	}

	$list_size = @whale;
	$notnew = 0;
	if ($list_size == 0) {
                push @pod, "1";
                push @counter, $number;
                push @whale, $key;
		$notnew = 1;
	}
	else {
		for ($i = 0; $i < $list_size; $i ++) {
			if ($whale[$i] eq $key) {
				$pod[$i] ++;
				$counter[$i] += $number;
				$notnew = 1;
				last;
			}
		}
	}
	
	if ($notnew == 0) {
		push @pod, "1";
		push @counter, $number;
		push @whale, $key;
	}
}

for ($i = 0; $i < @whale; $i ++) {
	$result = "$whale[$i] observations: $pod[$i] pods, $counter[$i] individuals\n";
	push @result_list, $result;
}

@sorted_result = sort @result_list;
for ($i = 0; $i < @whale; $i ++) {
	print $sorted_result[$i];
}

File: COMP9041/lab06/all_whales.py
#!/usr/bin/python2.7

import sys
import re

whale = []
pod = []
counter = []
result_list = []

for line in sys.stdin:
	line = re.sub(r" +", " ", line)
	line = re.sub(r"\n", "", line)
	line = re.sub(r" +$", "", line)
	line = line.lower()
	sp = line.split(" ", 1)

	number = int(sp[0])
	key = sp[1]
	if (re.match(r".+s$", key)):
		key = key[:-1]

	notnew = 0
	if (len(whale) == 0):
                whale.append(key)
		pod.append(1)
		counter.append(number)
		notnew = 1;
	else:
		for i in range(0, len(whale)):
			if (whale[i] == key):
				pod[i] += 1
				counter[i] += number
				notnew = 1
				break

	if (notnew == 0):
                whale.append(key)
		pod.append(1)
		counter.append(number)
	
for i in range(0, len(whale)):
	result = whale[i] + " observations: " + str(pod[i]) +" pods, " + str(counter[i]) + " individuals"
	result_list.append(result)

result_list.sort()
for i in range(0, len(whale)):
	print result_list[i]
